modules/imgcodecs/test/test_avif.cpp

```diff
@@ -0,0 +1,353 @@
+  }
+
+  // Makes sure images are close enough after encode/decode roundtrip.
+  void ValidateRead(const cv::Mat& img_original, const cv::Mat& img) const {
+    EXPECT_EQ(img_original.size(), img.size());
+    if (imread_mode_ == IMREAD_UNCHANGED) {
+      ASSERT_EQ(img_original.type(), img.type());
+      // Lossless.
+      if (quality_ == 100) {
+        EXPECT_EQ(0, cvtest::norm(img, img_original, NORM_INF));
+      } else {
+        const float norm = cvtest::norm(img, img_original, NORM_L2) /
+                           img.channels() / img.cols / img.rows /
+                           (1 << (bit_depth_ - 8));
+        if (quality_ == 50) {
+          EXPECT_LE(norm, 10);
+        } else if (quality_ == 0) {
+          EXPECT_LE(norm, 13);
+        } else {
+          EXPECT_FALSE(true);
+        }
+      }
+    }
+  }
+
+ public:
+  int bit_depth_;
+  int channels_;
+  int quality_;
+  int imread_mode_;
+  std::vector<int> encoding_params_;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+class Imgcodecs_Avif_Image_RoundTripSuite
+    : public Imgcodecs_Avif_RoundTripSuite {
+ public:
+  const cv::Mat& get_img_original() {
+    const Key key = {channels_, (bit_depth_ < 8) ? 8 : bit_depth_};
+    return imgs_[key];
+  }
+
+  // Prepare the original image modified for different number of channels and
+  // bit depth.
+  static void SetUpTestCase() {
+    const string root = cvtest::TS::ptr()->get_data_path();
+    const string filename = root + "../cv/shared/lena.png";
+    const cv::Mat img_original = cv::imread(filename);
+    cv::Mat img_resized;
+    cv::resize(img_original, img_resized, cv::Size(kWidth, kHeight), 0, 0);
+    for (int channels : {1, 3, 4}) {
+      for (int bit_depth : {8, 10, 12}) {
+        const Key key{channels, bit_depth};
+        imgs_[key] = modifyImage(img_resized, channels, bit_depth);
+      }
+    }
+  }
+
+  static const int kWidth;
+  static const int kHeight;
+
+ private:
+  typedef std::tuple<int, int> Key;
+  static std::map<Key, cv::Mat> imgs_;
+};
+std::map<std::tuple<int, int>, cv::Mat>
+    Imgcodecs_Avif_Image_RoundTripSuite::imgs_;
+const int Imgcodecs_Avif_Image_RoundTripSuite::kWidth = 51;
+const int Imgcodecs_Avif_Image_RoundTripSuite::kHeight = 51;
+
+class Imgcodecs_Avif_Image_WriteReadSuite
+    : public Imgcodecs_Avif_Image_RoundTripSuite {};
+
+TEST_P(Imgcodecs_Avif_Image_WriteReadSuite, imwrite_imread) {
+  const cv::Mat& img_original = get_img_original();
+  ASSERT_FALSE(img_original.empty());
+
+  // Encode.
+  const string output = cv::tempfile(".avif");
+  if (!IsBitDepthValid()) {
+    EXPECT_NO_FATAL_FAILURE(
+        cv::imwrite(output, img_original, encoding_params_));
+    return;
+  }
+  EXPECT_NO_THROW(cv::imwrite(output, img_original, encoding_params_));
+
+  // Read from file.
+  const cv::Mat img = cv::imread(output, imread_mode_);
+
+  ValidateRead(img_original, img);
```

Comment:
temporary file is not removed in the test.

Reply:
Fixed